### PR TITLE
WIP: experiment with line directives

### PR DESCRIPTION
**Pull request by gpetiot**

### CODE REVIEW

**Review by Julow — COMMENTED** [review]
I think that's a good idea :) Though, the formatting is not reliable due to the current implementation of `Fmt`.

I have no solution for the newline problem.

File: vendor/parser-extended/lexer.mll [hunk 1]
```diff
@@ -617,13 +617,14 @@ rule token = parse
       { error lexbuf (Illegal_character illegal_char) }
 
 and directive = parse
-  | ([' ' '\t']* (['0'-'9']+ as _num) [' ' '\t']*
-        ("\"" ([^ '\010' '\013' '\"' ] * as _name) "\"") as directive)
+  | ([' ' '\t']* (['0'-'9']+ as line) [' ' '\t']*
+        ("\"" ([^ '\010' '\013' '\"' ] * as name) "\""))
```
[the remark this answers is not in the document]
Comment (gpetiot) [reply vendor/parser-extended/lexer.mll:621]: I don't think cppo directives can be handled in the same way. I was thinking that we can encode the conditions as regular if expressions with an attribute indicating they are cppo conditions, but didn't give it more thought than that.

**Comment by hhugo** [comment]
Can you explain the expected workflow for formatting files containing line directive ? It's weird to try to preserve location with line directive and also shuffle location while formatting.

**Comment by gpetiot** [comment]
> Can you explain the expected workflow for formatting files containing line directive ? It's weird to try to preserve location with line directive and also shuffle location while formatting.

It gives the users the choice to format or not their file, they can still list their files in `.ocamlformat-ignore`. It makes sense to me that users would want to format a file containing a line directive and then manually edit the line directive if they wish to get something closer to the new shape of the file.

**Comment by hhugo** [comment]
> > Can you explain the expected workflow for formatting files containing line directive ? It's weird to try to preserve location with line directive and also shuffle location while formatting.
> 
> It gives the users the choice to format or not their file, they can still list their files in `.ocamlformat-ignore`. It makes sense to me that users would want to format a file containing a line directive and then manually edit the line directive if they wish to get something closer to the new shape of the file.

The only workflow I know for line directive is inside automatically generated files in which case the source files are not meant to be edited directly. Changing the shape of a file with line directive means that locations updated by the line directive are now incorrect and useless.

**Comment by panglesd** [comment]
I had a vague idea on how to use that to format cppo-processed files, at least some of them:
- cppo-process them to get a file with line-directives
- ocamlformat this file with line directives
- Use the line directives to reconstruct a formatted original file. Some thing along the lines of:
  - Separate the each file in blocks separated by directives (`#if`, `#else`, ...)
  - Match the blocks using the line directives: if two block start at the same line they match
  - Replace the unformatted blocks by formatted one.

This is just the beginning of an idea and that might be a bad one, there might be simpler ways to format cppo files, but that's what triggered my interest in formatting line directives (which, compared to cppo directives such as `#if`, are at least valid OCaml).

**Comment by hhugo** [comment]
> I had a vague idea on how to use that to format cppo-processed files, at least some of them:
> 
> * cppo-process them to get a file with line-directives
> * ocamlformat this file with line directives
> * Use the line directives to reconstruct a formatted original file. Something along the lines of:
>   
>   * Separate each file in blocks separated by directives (`#if`, `#else`, ...)
>   * Match the blocks using the line directives: if two block start at the same line they match
>   * Replace the unformatted blocks by formatted one.
> 
> This is just the beginning of an idea and that might be a bad one, there might be simpler ways to format cppo files, but that's what triggered my interest in formatting line directives (which, compared to cppo directives such as `#if`, are at least valid OCaml).

This seems to be a very convoluted workflow that gives many opportunities for failures.
There exists an alternative to cppo which works fine with ocamlformat already (and maybe merlin as well to some degree).
See https://github.com/janestreet/ppx_optcomp
It's not as powerful as cppo but could be extended to fit the need.

I use a custom ppx in js_of_ocaml that uses the same principals.
See https://github.com/ocsigen/js_of_ocaml/blob/master/compiler/ppx/ppx_optcomp_light.ml

I suspect efforts would be better spent extending (or replacing) ppx_optcomp.

**Pull request closed without merging**